test/common/http/http2/codec_impl_test.cc

```diff
@@ -655,6 +655,158 @@ TEST_P(Http2CodecImplTest, ShutdownNotice) {
+  ASSERT_EQ(0, server_stats_store_.counter("http2.goaway_graceful_sent").value());
+  ASSERT_EQ(0, server_stats_store_.counter("http2.goaway_sent").value());
+
+  // Allow GOAWAY callbacks to be called (these happen automatically when frames are processed)
+  EXPECT_CALL(client_callbacks_, onGoAway(_)).Times(AtLeast(1));
+
+  server_->goAwayGraceful();
+  driveToCompletion();
+
+  // Verify graceful GOAWAY was sent (stats should increment)
+  EXPECT_EQ(1, server_stats_store_.counter("http2.goaway_graceful_sent").value());
+  EXPECT_EQ(0, server_stats_store_.counter("http2.goaway_sent").value());
+
+  // Trigger timeout and final GOAWAY
+  graceful_goaway_timer->invokeCallback();
+  driveToCompletion();
+
+  // Verify final GOAWAY was sent
+  EXPECT_EQ(1, server_stats_store_.counter("http2.goaway_graceful_sent").value());
+  EXPECT_EQ(1, server_stats_store_.counter("http2.goaway_sent").value());
+}
+
+TEST_P(Http2CodecImplTest, GracefulGoAwayFallbackWhenTimeoutZero) {
+  // Configure graceful GOAWAY timeout of 0 (disabled)
+  server_http2_options_.mutable_graceful_goaway_timeout()->set_seconds(0);
+  server_http2_options_.mutable_graceful_goaway_timeout()->set_nanos(0);
+  initialize();
+
+  // Should fallback to immediate GOAWAY
+  ASSERT_EQ(0, server_stats_store_.counter("http2.goaway_graceful_sent").value());
+  ASSERT_EQ(0, server_stats_store_.counter("http2.goaway_sent").value());
+
+  EXPECT_CALL(client_callbacks_, onGoAway(_)).Times(AtLeast(1));
+  server_->goAwayGraceful();
+  driveToCompletion();
+
+  // Verify immediate GOAWAY was sent (no graceful GOAWAY)
+  EXPECT_EQ(0, server_stats_store_.counter("http2.goaway_graceful_sent").value());
+  EXPECT_EQ(1, server_stats_store_.counter("http2.goaway_sent").value());
+}
+
+TEST_P(Http2CodecImplTest, GracefulGoAwayAlreadyInProgress) {
+  // Configure graceful GOAWAY timeout of 100ms
+  server_http2_options_.mutable_graceful_goaway_timeout()->set_seconds(0);
+  server_http2_options_.mutable_graceful_goaway_timeout()->set_nanos(100 * 1000 * 1000); // 100ms
+
+  auto graceful_goaway_timer = new NiceMock<Event::MockTimer>(&server_connection_.dispatcher_);
+  EXPECT_CALL(*graceful_goaway_timer, enableTimer(std::chrono::milliseconds(100), _));
+
+  initialize();
+
+  // First graceful GOAWAY call
+  EXPECT_CALL(client_callbacks_, onGoAway(_)).Times(AtLeast(1));
+  server_->goAwayGraceful();
+  driveToCompletion();
+
+  EXPECT_EQ(1, server_stats_store_.counter("http2.goaway_graceful_sent").value());
+  EXPECT_EQ(0, server_stats_store_.counter("http2.goaway_sent").value());
+
+  // Second graceful GOAWAY call should fallback to immediate GOAWAY
+  server_->goAwayGraceful();
+  driveToCompletion();
+
+  // Should have sent immediate GOAWAY
+  EXPECT_EQ(1, server_stats_store_.counter("http2.goaway_graceful_sent").value());
+  EXPECT_EQ(1, server_stats_store_.counter("http2.goaway_sent").value());
+}
+
+TEST_P(Http2CodecImplTest, GracefulGoAwayCausesOutboundFlood) {
```

Comment:
I think it would be prudent to add some tests showing how this new feature and the existing http connection manager drain phase interact.

Could you also add a couple integration tests?

Reply:
TY for the added testing!

I think what worries me the most is if it's possible for the drain timer to fire after the graceful shutdown timer or vice versa.

Will that crash the envoy? (e.g. the second timer fires and tries calling a function on a now null object).

Could that potentially cause us to send two final goaways (with the second having a higher stream_id-- which is not allowed according to the spec)?

Reply:
As far as I understand the code (which is little to be honest) -- the destructor on `ConnectionImpl` should destroy the timers -> cancelling any pending callbacks -> preventing null object access.

Actually the behavior hasn't changed much. I reverted the change on calling `gracefulGoAway` on `onDrainTimeout`. Now only filter or load shedding invoked goaways are changed to graceful. Both goaway paths are separate and non overlapping now. It just changes the manual goaway to be more RFC compliant.

However, I will keep attempting adding test.